File: src/backend/loop_diagnostics.py
"""Bounded, content-free observations of a LOOP model call (not provider guesses)."""
from __future__ import annotations

import asyncio
import copy
import re
import time
from typing import Callable

from .token_usage import estimate_tokens


def error_evidence(error: object) -> dict:
    # 不保存原始异常：SDK 报错可能回显请求、密钥或带认证的 URL。
    text = str(error).lower()
    status = re.search(r"\b([45]\d{2})\b", text)
    category = "unknown"
    if re.search(r"concurren|并发", text):
        category = "concurrency"
    elif (status and status[1] == "429") or re.search(r"rate.?limit|限流|too many requests", text):
        category = "rate_limit"
    elif re.search(r"timeout|timed out|超时", text):
        category = "timeout"
    elif re.search(r"connect|network|protocol|网络|连接", text):
        category = "network"
    elif status and status[1] in ("401", "403"):
        category = "auth"
    elif re.search(r"context.?length|too many tokens|上下文.*超", text):
        category = "context_limit"
    return {"category": category, **({"httpStatus": int(status[1])} if status else {})}


class LoopCallDiagnostics:
    """One mutable record in the loop cache; pushes ≤1/s, no idle polling."""
    def __init__(self, call_id: str, stage: str, prompt: str,
                 publish: Callable[[dict], None], persist: Callable[[], None]) -> None:
        self._publish = publish
        self._persist = persist
        self._timer: asyncio.TimerHandle | None = None
        self._last_push = self._last_save = 0.0
        self._mono = time.monotonic()
        self.closed = False
        self.data: dict = {
            "id": call_id, "stage": stage, "status": "running", "phase": "local_prepare",
            "startedAt": time.time(), "promptChars": len(prompt),
            "estimatedPromptTokens": estimate_tokens(prompt), "imageCount": 0,
            "eventCounts": {}, "textChars": 0, "thinkingChars": 0,
            "transportAttempts": 0, "retryCount": 0, "retryWaitSeconds": 0,
            "timeline": [],
        }

    def mark(self, phase: str, **fields: object) -> None:
        if self.closed:
            return
        self.data.update(fields)
        if self.data["phase"] != phase or not self.data["timeline"]:
            self.data["phase"] = phase
            self.data["timeline"].append({"phase": phase, "at": time.time(),
                "elapsedMs": round((time.monotonic() - self._mono) * 1000),
                **{k: fields[k] for k in ("httpStatus", "attempt", "delaySeconds", "category") if k in fields}})
            self.data["timeline"] = self.data["timeline"][-32:]
        self.changed()

    def changed(self) -> None:
        if self.closed:
            return
        if time.monotonic() - self._last_push >= 1:
            self.flush()
        elif self._timer is None:
            self._timer = asyncio.get_running_loop().call_later(1, self.flush)

    def flush(self) -> None:
        if self._timer:
            self._timer.cancel()
            self._timer = None
        now = time.monotonic()
        self._last_push = now
        self.data["observedAt"] = time.time()
        self._publish(copy.deepcopy(self.data))
        if self.closed or now - self._last_save >= 10:
            self._last_save = now
            self._persist()

    def observe(self, delta: object) -> None:
        if self.closed:
            return
        kind = getattr(delta, "type", "")
        if kind == "diagnostic":
            meta = getattr(delta, "diagnostic", None) or {}
            phase = meta.get("phase")
            if phase not in ("request", "response_headers", "retry_wait", "runner_start", "runner_ready", "turn_accepted", "backend_error", "sdk_request", "sdk_response"):
                return
            # 白名单仅接受数字和固定分类，不传播任意 Backend 数据。
            fields = {k: v for k, v in meta.items() if k in (
                "httpStatus", "attempt", "maxAttempts", "delaySeconds", "requestBytes"
            ) and isinstance(v, (int, float)) and 0 <= v < 1e12}
            category = meta.get("category")
            if category in ("rate_limit", "network", "timeout", "concurrency", "auth", "context_limit", "unknown"):
                fields["category"] = category
                self.data["lastError"] = {"category": category}
            if fields.get("httpStatus", 0) >= 400:
                self.data["lastError"] = error_evidence(f"HTTP {fields['httpStatus']}")
            if phase == "request":
                self.data["transportAttempts"] += 1
            if phase == "retry_wait":
                self.data["retryCount"] += 1
                self.data["retryWaitSeconds"] += fields.get("delaySeconds", 0)
            if phase == "backend_error":
                self.data["hadError"] = True
            if isinstance(meta.get("model"), str):
                # 模型名是明确的配置字段；不接受任意状态原文或 URL。
                if re.fullmatch(r"[\w./:@+ -]{1,120}", meta["model"]):
                    self.data["model"] = meta["model"]
            self.mark(phase, **fields)
            return
        if kind not in ("text_delta", "thinking", "tool_start", "tool_input", "tool_result",
                         "subagent_start", "subagent_progress", "subagent_done", "error", "done"):
            return
        now = time.time()
        self.data.setdefault("firstEventAt", now)
        self.data["lastActivityAt"] = now
        counts = self.data["eventCounts"]
        counts[kind] = counts.get(kind, 0) + 1
        if kind == "done":
            usage = getattr(delta, "usage", None) or {}
            counters = {k: int(v) for k, v in usage.items()
                if k in ("inputTokens", "outputTokens", "cachedInputTokens", "reasoningOutputTokens")
                and isinstance(v, (int, float)) and 0 <= v < 1e15}
            if counters:
                self.data["usage"] = counters
        text = getattr(delta, "text", None) or ""
        if kind == "error":
            self.fail(getattr(delta, "error", ""))
        elif kind == "text_delta" and text:
            self.data.setdefault("firstTextAt", now)
            self.data["textChars"] += len(text)
            self.mark("text")
        elif kind == "thinking" and text:
            self.data["thinkingChars"] += len(text)
            self.mark("thinking")
        elif kind.startswith("tool_") or kind.startswith("subagent_"):
            self.mark("tools")
        self.changed()

    def fail(self, error: object) -> None:
        if not self.closed:
            self.data["hadError"] = True
            self.data["lastError"] = error_evidence(error)
            self.mark("error")

    def finish(self, status: str = "done") -> None:
        if self.closed:
            return
        if status == "done" and self.data.get("hadError"):
            status = "error"
        self.mark(status, status=status, endedAt=time.time(),
                  durationMs=round((time.monotonic() - self._mono) * 1000))
        self.closed = True
        self.flush()

File: frontend/src/utils/loopDiagnostics.ts
export interface CallDiagnostic {
  id: string; stage: string; status: string; phase: string;
  startedAt: number; dispatchedAt?: number; endedAt?: number; observedAt?: number;
  firstEventAt?: number; firstTextAt?: number; lastActivityAt?: number;
  localPrepareMs?: number; durationMs?: number;
  backendId?: string; backendType?: string; model?: string; reasoningEffort?: string;
  promptChars: number; estimatedPromptTokens: number; imageCount: number; resumedContext?: boolean;
  requestBytes?: number; inactivityTimeoutSeconds?: number;
  activeLoopCallsAtDispatch?: number;
  usage?: { inputTokens?: number; outputTokens?: number; cachedInputTokens?: number; reasoningOutputTokens?: number };
  textChars: number; thinkingChars: number; eventCounts: Record<string, number>;
  transportAttempts: number; retryCount: number; retryWaitSeconds: number;
  httpStatus?: number; attempt?: number; maxAttempts?: number; delaySeconds?: number;
  lastError?: { category: string; httpStatus?: number };
  timeline?: Array<{ phase: string; at: number; elapsedMs: number; httpStatus?: number; attempt?: number; delaySeconds?: number; category?: string }>;
}

export const CALL_PHASE: Record<string, string> = {
  local_prepare: '本地准备：同步 Skill／展开引用', backend_dispatch: '已进入 Backend，等待可见事件',
  running: '进行中',
  runner_start: '启动 Agent 进程／连接', runner_ready: 'Agent 就绪，初始化线程',
  turn_accepted: 'Agent 已接受任务，等待模型事件', request: '已发起 HTTP 请求，等待响应头',
  response_headers: '已收到 HTTP 响应头', retry_wait: '后端自动重试等待',
  thinking: '收到思考事件，尚不等于计划完成', text: '收到正文', tools: '工具／子任务活动',
  done: '调用结束（不代表计划校验通过）', error: '调用报错', stalled: '无事件超时', cancelled: '调用已中断',
  backend_error: 'Agent／SDK 报错',
  sdk_request: '已交给 API SDK，内部排队／重试不可见', sdk_response: 'API SDK 已建立响应流',
};
export const CALL_ERROR: Record<string, string> = {
  concurrency: '后端明确报告并发限制', rate_limit: '限流（不一定是并发，也可能是 RPM／TPM）',
  timeout: '超时', network: '网络／连接错误', auth: '认证／权限错误',
  context_limit: '上下文长度限制', unknown: '未分类错误；不能据此推断并发不足',
};

export function mergeCallDiagnostics(...groups: Array<CallDiagnostic[] | undefined>): CallDiagnostic[] {
  const calls = new Map<string, CallDiagnostic>();
  for (const group of groups) for (const call of group || []) {
    const previous = calls.get(call.id);
    if (!previous || (call.endedAt || call.observedAt || call.startedAt) >= (previous.endedAt || previous.observedAt || previous.startedAt)) {
      calls.set(call.id, call);
    }
  }
  return [...calls.values()].slice(-64);
}

// 只有本轮最后一条未终止记录才可能在跑。旧轮次残留 running 不能抢占动画。
export function activeLoopSeq(state: { running: boolean; round: number; loops: Array<{
  seq: number; round: number; completed: boolean; error: string; kind?: string;
}> }): number | null {
  const last = state.loops[state.loops.length - 1];
  return state.running && last && last.round === state.round && last.kind !== 'manual'
    && !last.completed && !last.error ? last.seq : null;
}
